### PR TITLE
join using page_op in memory

### PR DESCRIPTION
**Pull request by abuts**
This is part of ticket #1402 which implements join using PageOp in case when all input and output data are located in memory.

### CODE REVIEW

File: horace_core/sqw/@Experiment/Experiment.m [hunk 1]
```diff
@@ -259,6 +259,13 @@
         function obj = set.runid_recalculated(obj,val)
             % Do not normally use, except for tests. This is internal
             % property, which inform about the behaviour of runid map
+            % Inputs:
+            % val  -- either containers.Map which define correspondence
+            %         between run_id and header index or array of unique 
+            %         numbers which will become new run_id. 
+            %
+            %         Syncrhonization with run_id stored in pixels have to
+            %         be performed separately.
```
Comment (oerc0122) [inline horace_core/sqw/@Experiment/Experiment.m:268]:
```suggestion
            % val  -- either:
            %         - containers.Map which defines correspondence
            %         between run_id and header index 
            %         OR 
            %         - array of unique numbers which will become new run_id. 
            %
            %         Synchronization with run_id stored in pixels has to
            %         be performed separately.
```

File: horace_core/sqw/PixelData/@pixobj_combine_info/pixobj_combine_info.m [hunk 2]
```diff
@@ -55,32 +52,80 @@
...
+            for i=1:numel(obj.infiles_)
+                obj.infiles_{i}.keep_precision = true;
+            end
+        end
+
+        function [data,npix_chunk] = get_dataset_page(obj, ...
+                n_dataset,pix_pos,npix_idx)
+            % Return pixel data and pixel bin sub-distribution for the
+            % particular dataset ouf of multiple pixel datasets, stored
+            % within the class.
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@pixobj_combine_info/pixobj_combine_info.m:85]: I can't work out what this function does. Is this right: 

It takes a contiguous slice of pixels of a given size from the underlying `sqw` objects even if it covers multiple `sqw` objects


```suggestion
            % Return pixel data and pixel bin sub-distribution for the
            % particular dataset out of multiple pixel datasets, stored
            % within the class.
```

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/check_combo_arg_.m [hunk 3]
```diff
@@ -0,0 +1,23 @@
+function obj = check_combo_arg_(obj)
+%CHECK_COMBO_ARG_ validate consistency of cellarray of pixels data and
+% cellarray of distributions, describing these pixels
+if isempty(obj.npix_list_) && ~isempty(obj.infiles_)
+    obj.do_check_combo_arg_ = false; % avoid calling this routine recursively
+    obj.npix_list = num2cell(obj.npix_each_file_);
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@pixobj_combine_info/private/check_combo_arg_.m:6]: Why does a `check` routine change things?

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m [hunk 4]
```diff
@@ -0,0 +1,25 @@
+function obj = set_npix_list_(obj,val)
+% Main setter for list of npix arrays, describing distrinution of pixels
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m:2]:
```suggestion
% Main setter for list of npix arrays, describing distribution of pixels
```

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+% over bins
+% Accepts single npix or cellarry of npix distributions. All heed to have
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m:4]:
```suggestion
% Accepts single npix or cellarry of npix distributions. All need to have
```

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+% the same number of elements.
+
+if iscell(val)
+    nbins = numel(val{1});
+    eq_nbins = cellfun(@(x)(numel(x)==nbins),val);
+    if ~all(eq_nbins)
+        error('HORACE:pixobj_combine_info:invalid_argument', ...
+            ['All pixels distributions (npix) should contain equal number of bins.\n' ...
+            ' There ara %d distributions which does not'],sum(~eq_nbins));
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m:13]:
```suggestion
            ' There are %d distributions which does not'],sum(~eq_nbins));
```

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    end
+    val_norm = cellfun(@(x)x(:),val,'UniformOutput',false);
+    obj.npix_list_ = val_norm(:);
+
+elseif isnumeric(val)
+    obj.npix_list_ = {val(:)};
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m:19]: So if it's numeric there are no checks on validity is this correct?

File: horace_core/sqw/page_operations/@PageOpBase/private/init_.m [hunk 8]
```diff
@@ -25,7 +25,7 @@
         class(in_obj))
 end
 % as we normally read data and immediately dump them back, what
-% is the point of converting them to double and back?
+% is the point of converting them to double and back to sinlge?
```
Comment (oerc0122) [inline horace_core/sqw/page_operations/@PageOpBase/private/init_.m:28]:
```suggestion
% is the point of converting them to double and back to single?
```

File: horace_core/sqw/page_operations/PageOp_join_sqw.m [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            obj.new_runid = new_runid;
+            % clear signal accumulator to save memory; it will not be used
+            % here.
+            obj.sig_acc_  = [];
+            % initialize input datasets for read access
+            obj.pix_combine_info  = obj.pix_combine_info.init_pix_access();
+            obj.current_page_pix_pos_ = ones(1,obj.pix_combine_info.nfiles);
+        end
+        function [npix_chunks, npix_idx,obj] = split_into_pages(obj,npix,chunk_size)
+            % Overload of split method allowing to define large target chink
```
Comment (oerc0122) [inline horace_core/sqw/page_operations/PageOp_join_sqw.m:56]:
```suggestion
            % Overload of split method allowing to define large target chunk
```

File: horace_core/sqw/page_operations/PageOp_join_sqw.m [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            %                the npix array.
+            % See split procedure for more details
+            fb = config_store.instance().get_value( ...
+                'hor_config','fb_scale_factor');
+            % do large chunk to decrease number of sub-calls to each data
+            % pixels
+            large_chunk = chunk_size*fb;
+            [npix_chunks, npix_idx,obj] = split_into_pages@PageOpBase(obj,npix,large_chunk);
+            obj.block_idx_ = npix_idx;
+            page_sizes = cellfun(@(x)sum(x(:)),npix_chunks);
```
Comment (oerc0122) [inline horace_core/sqw/page_operations/PageOp_join_sqw.m:76]: Could just be
```suggestion
            page_sizes = cellfun(@sum, npix_chunks);
```

File: horace_core/sqw/page_operations/PageOp_join_sqw.m [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                % advance initial page position for the particular dataset
+                n_page_pix = size(contr_page_data,2);
+                obj.current_page_pix_pos_(i) = ...
+                    obj.current_page_pix_pos_(i)+n_page_pix;
+                %
+                if ~isempty(obj.new_runid)
+                    contr_page_data(obj.run_idx_,:) = obj.new_runid(i);
+                end
+
+                % find indexes of i-th dataset's page pixels in the target's
```
Comment (oerc0122) [inline horace_core/sqw/page_operations/PageOp_join_sqw.m:107]:
```suggestion
                % find indices of i-th dataset's page pixels in the target's
```

File: horace_core/sqw/page_operations/PageOp_join_sqw.m [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+                % dataset page
+                targ_bin_idx = fill_idx(bin_start,page_bin_distr);
+                % targ_bin_pos   = repelem(bin_start,page_bin_distr);
+                % cell_idx = arrayfun(@fill_fun,page_bin_distr,'UniformOutput',false);
+                % targ_bin_idx   = targ_bin_pos+[cell_idx{:}]';
```
Comment (oerc0122) [inline horace_core/sqw/page_operations/PageOp_join_sqw.m:112]: Debugging statements?

File: horace_core/sqw/page_operations/@PageOpBase/private/init_.m (hunk 8 above)
Comment (oerc0122) [reply horace_core/sqw/page_operations/@PageOpBase/private/init_.m:28]: Though I'm not sure of the need for this, nor that it should be in this PR.

File: _test/test_sqw/test_pixobj_combine_info.m [hunk 13]
```diff
@@ -2,7 +2,8 @@
     % Series of tests to check validity of pixobj_combine_info class
 
     properties
-        npix_tst = [10,15,20]
+        npix_tst = [10,16,20]
+        npix_dirst
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_sqw/test_pixobj_combine_info.m:6]: distr indeed.

File: horace_core/algorithms/check_img_consistency.m [hunk 14]
```diff
@@ -0,0 +1,67 @@
+function [ok,mess,img_range] = check_img_consistency(sqw_obj_input,throw_on_inconsistent)
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/algorithms/check_img_consistency.m:1]: probably OK, but I will fix it with PR #1425 as routine have substantially changed

File: horace_core/sqw/PixelData/@MultipixBase/MultipixBase.m [hunk 15]
```diff
@@ -117,6 +125,12 @@
             obj.do_check_combo_arg_= true;
         end
         %------------------------------------------------------------------
+        function obj = init_pix_access(obj)
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/PixelData/@MultipixBase/MultipixBase.m:128]: May be I've done this in the following modifications as both are used and different

[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/PixelData/@MultipixBase/MultipixBase.m:128]: ok, have not, can be done here.

File: horace_core/sqw/PixelData/@PixelDataBase/PixelDataBase.m [hunk 16]
```diff
@@ -139,8 +139,8 @@
         % memory or on disk and all additional properties describing
         % pix array, like its size, shape, alignment, etc
         metadata;
-        % the property contains or describes the pixel data array itself
-        % contains if the array fits area or describes it if the only
+        % the property contains or describes the pixel data array itself.
+        % Contains if the array fits area or describes it if the only
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/PixelData/@PixelDataBase/PixelDataBase.m:143]: I was probably using it somewhere.

File: horace_core/sqw/@Experiment/Experiment.m (hunk 1 above)
Comment (abuts) [reply horace_core/sqw/@Experiment/Experiment.m:268]: Thsis is absolutely wrong and confusing comment. Removing it.

File: horace_core/sqw/PixelData/@pixobj_combine_info/pixobj_combine_info.m (hunk 2 above)
Comment (abuts) [reply horace_core/sqw/PixelData/@pixobj_combine_info/pixobj_combine_info.m:85]: its `pixobj_combine` object, which contains information about  pixels and images (pat of image -- `npix`) of all combining datasets. This method takes infor from one dataset. No `sqw` objects

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/check_combo_arg_.m (hunk 3 above)
Comment (abuts) [reply horace_core/sqw/PixelData/@pixobj_combine_info/private/check_combo_arg_.m:6]: well -- may be it should be renamed. But it changes it in multiple places.

File: horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m (hunk 7 above)
Comment (abuts) [reply horace_core/sqw/PixelData/@pixobj_combine_info/private/set_npix_list_.m:19]: further checks agains possible human error are performed in `check_combo_arg` but as the inputs are coming from presimably valid sqw objects, no thorough checks are necessary.

File: horace_core/sqw/page_operations/PageOp_join_sqw.m (hunk 10 above)
Comment (abuts) [reply horace_core/sqw/page_operations/PageOp_join_sqw.m:76]: are you sure chunks have correct shape (1D) ? . Better to be safe then sorry.

File: horace_core/sqw/page_operations/PageOp_join_sqw.m (hunk 12 above)
Comment (abuts) [reply horace_core/sqw/page_operations/PageOp_join_sqw.m:112]: Will be checked tomorrow. Letf for profiling

**Pull request merged**